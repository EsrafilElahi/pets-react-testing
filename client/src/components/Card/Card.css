.card {
  width: 16rem;
  height: 22.5rem;
  text-align: center;
  color: rgb(97, 8, 97);
  margin-right: 2rem;
  margin-bottom: 2rem;
  box-shadow: 0.1rem 0.1rem 1rem rgba(0, 0, 0, 0.486);
}

.card-header {
  padding: 0;
  position: relative;
  height: 70%;
}

.heart {
  border-radius: 100%;
  background-color: rgba(255, 255, 255, 0.678);
  width: 3rem;
  height: 3rem;
  z-index: 100;
  position: absolute;
  bottom: 78%;
  left: 78%;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  border: none;
}

.heart img {
  width: 1.75rem;
}

.card-img {
  width: 100%;
  height: 100%;
}

.card h3 {
  margin: 0.5rem 0;
}

.card p {
  margin: 0;
}
