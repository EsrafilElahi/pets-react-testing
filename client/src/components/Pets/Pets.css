.app-container {
  display: flex;
  padding: 3rem 0;
}
