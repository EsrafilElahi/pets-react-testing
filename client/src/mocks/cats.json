[
  {
    "id": 1,
    "name": "Sydney",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1529778873920-4da4926a72c2?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MTJ8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "female"
  },
  {
    "id": 2,
    "name": "Ja",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1536590158209-e9d615d525e4?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MTl8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "male"
  },
  {
    "id": 3,
    "name": "Holly",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1606214174585-fe31582dc6ee?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MjB8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "female"
  },
  {
    "id": 4,
    "name": "Kevin",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1592194996308-7b43878e84a6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8OXx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "male"
  },
  {
    "id": 5,
    "name": "Rachel",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1561948955-570b270e7c36?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8NXx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "female"
  }
]
