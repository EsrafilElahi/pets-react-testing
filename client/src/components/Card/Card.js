import { useContext, useState } from "react";
import heartFilled from "../../svgs/heartFilled.svg";
import heartOutlined from "../../svgs/heartOutlined.svg";
import { PetsContext } from "../Pets/Pets";
import "./Card.css";

const Card = ({ name, phone, email, image, favoured, index }) => {
  const { cats, setCats } = useContext(PetsContext);
  const [isFavoured, setIsFavoured] = useState(favoured);

  const updateFavourite = (index, favoured) => {
    const updatedCats = [...cats];
    updatedCats[index].favoured = favoured;
    setCats(updatedCats);
  };

  const toggleFavoured = () => {
    updateFavourite(index, !isFavoured);
    setIsFavoured(!isFavoured);
  };

  return (
    <article className="card">
      <div className="card-header">
        <img src={image.url} alt={image.alt} className="card-img" />
        <button className="heart" onClick={toggleFavoured}>
          {isFavoured ? (
            <img src={heartFilled} alt="filled heart" />
          ) : (
            <img src={heartOutlined} alt="outlined heart" />
          )}
        </button>
      </div>
      <div className="card-content">
        <h3>{name}</h3>
        <p>{phone}</p>
        <p>{email}</p>
      </div>
    </article>
  );
};

export default Card;
