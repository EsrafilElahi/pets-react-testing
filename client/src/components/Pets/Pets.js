import "./Pets.css";
import Filter from "../Filter/Filter";
import Cards from "../Cards/Cards";

import axios from "axios";
import { useState, useEffect, createContext } from "react";

export const PetsContext = createContext({
  cats: [],
  setCats: () => {},
});

const Pets = () => {
  const [cats, setCats] = useState([]);
  const [filteredCats, setFilteredCats] = useState([]);
  const [filters, setFilters] = useState({
    gender: "any",
    favoured: "any",
  });

  const fetchCats = async () => {
    const response = await axios.get("http://localhost:4000/cats");
    setCats(response.data);
    setFilteredCats(response.data);
  };

  useEffect(() => {
    fetchCats();
  }, []);

  useEffect(() => {
    let catsFiltered = [...cats];
    if (filters.gender !== "any") {
      catsFiltered = catsFiltered.filter(
        (cat) => cat.gender === filters.gender
      );
    }
    if (filters.favoured !== "any") {
      catsFiltered = catsFiltered.filter((cat) => {
        console.log(
          cat.favoured,
          filters.favoured,
          cat.favoured === filters.favoured
        );
        return (
          cat.favoured === (filters.favoured === "favoured" ? true : false)
        );
      });
    }
    setFilteredCats(catsFiltered);
  }, [filters]);

  return (
    <div className="container">
      <div className="app-container">
        <PetsContext.Provider value={{ cats: filteredCats, setCats }}>
          <Filter filters={filters} setFilters={setFilters} />
          <Cards />
        </PetsContext.Provider>
      </div>
    </div>
  );
};

export default Pets;
