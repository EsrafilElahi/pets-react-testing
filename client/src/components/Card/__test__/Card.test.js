import { screen, render, getByRole } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Card from '../Card';
import cats from '../../../mocks/cats.json';
import { PetsContext } from '../../Pets/Pets';


const cardProps = {
  name: "Sydney",
  phone: "111-111-1111",
  email: "laith@hotmail.com",
  image: {
    url: "https://images.unsplash.com/photo-1548247416-ec66f4900b2e?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8Mnx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=1000&q=60",
    alt: "cute cat",
  },
  favoured: false,
  updateFavourite: () => { },
  index: 1,
};

const renderCardComponentWithProvider = (props) => {
  render(
    <PetsContext.Provider
      value={{
        cats: cats,
        setCats: () => { }
      }}
    >
      <Card {...props} />
    </PetsContext.Provider>
  );
};

describe('Card', () => {
  test("snapshot test", () => {
    const { asFragment } = render(
      <PetsContext.Provider
        value={{
          cats: cats,
          setCats: () => { }
        }}
      >
        <Card {...cardProps} />
      </PetsContext.Provider>
    );;
    expect(asFragment()).toMatchSnapshot();
  });

  describe("displays card details", () => {
    beforeEach(() => {
      return renderCardComponentWithProvider(cardProps);
    });

    test("name display in page", () => {
      // find card element
      const nameText = screen.getByRole("heading", { name: /sydney/i });
      // assertion
      expect(nameText).toBeInTheDocument();
    });

    test("phone-number display in page", () => {
      const phoneText = screen.getByText(/111-111-1111/i);
      expect(phoneText).toBeInTheDocument()
    });

    test("email display in page", () => {
      const emailText = screen.getByText(/laith@hotmail.com/i);
      expect(emailText).toBeInTheDocument();
    });

    test("image url display in page", () => {
      const imgElement = screen.getByAltText(/cute cat/i);
      expect(imgElement.src).toBe(cardProps.image.url);
    });
  });

  describe('like, dislike status', () => {

    test("outlined heart display in page", () => {
      renderCardComponentWithProvider(cardProps);

      const outlinedHeart = screen.getByAltText(/outlined heart/i);
      const filledHeart = screen.queryByAltText(/filled heart/i);

      expect(outlinedHeart).toBeInTheDocument();
      expect(filledHeart).not.toBeInTheDocument();
    });

    test("filled heart display in page", () => {
      renderCardComponentWithProvider({ ...cardProps, favoured: true });

      const filledHeart = screen.getByAltText(/filled heart/i);
      const outlinedHeart = screen.queryByAltText(/outlined heart/i);

      expect(filledHeart).toBeInTheDocument();
      expect(outlinedHeart).not.toBeInTheDocument();
    });

    test("toggle like", () => {
      renderCardComponentWithProvider(cardProps);

      const toggleBtn = screen.getByRole("button");

      // click 1 - like
      userEvent.click(toggleBtn);

      // assertion two items
      const filledHeart = screen.getByAltText(/filled heart/i);
      const outlinedHeart = screen.queryByAltText(/outlined heart/i);
      expect(filledHeart).toBeInTheDocument();
      expect(outlinedHeart).not.toBeInTheDocument();

      // click 2 - dislike
      userEvent.click(toggleBtn);

      // assertion two items
      const outlinedHeart2 = screen.getByAltText(/outlined heart/i);
      const filledHeart2 = screen.queryByAltText(/filled heart/i);

      expect(outlinedHeart2).toBeInTheDocument();
      expect(filledHeart2).not.toBeInTheDocument();
    });
  });

})
