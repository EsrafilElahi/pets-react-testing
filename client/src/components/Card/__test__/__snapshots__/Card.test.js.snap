// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card displays card details snapshot test 1`] = `
<DocumentFragment>
  <article
    class="card"
  >
    <div
      class="card-header"
    >
      <img
        alt="cute cat"
        class="card-img"
        src="https://images.unsplash.com/photo-1548247416-ec66f4900b2e?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8Mnx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=1000&q=60"
      />
      <button
        class="heart"
      >
        <img
          alt="outlined heart"
          src="heartOutlined.svg"
        />
      </button>
    </div>
    <div
      class="card-content"
    >
      <h3>
        Sydney
      </h3>
      <p>
        111-111-1111
      </p>
      <p>
        laith@hotmail.com
      </p>
      <p />
    </div>
  </article>
</DocumentFragment>
`;

exports[`Card like, dislike status snapshot test 1`] = `
<DocumentFragment>
  <article
    class="card"
  >
    <div
      class="card-header"
    >
      <img
        alt="cute cat"
        class="card-img"
        src="https://images.unsplash.com/photo-1548247416-ec66f4900b2e?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8Mnx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=1000&q=60"
      />
      <button
        class="heart"
      >
        <img
          alt="outlined heart"
          src="heartOutlined.svg"
        />
      </button>
    </div>
    <div
      class="card-content"
    >
      <h3>
        Sydney
      </h3>
      <p>
        111-111-1111
      </p>
      <p>
        laith@hotmail.com
      </p>
      <p />
    </div>
  </article>
</DocumentFragment>
`;

exports[`Card snapshot test 1`] = `
<DocumentFragment>
  <article
    class="card"
  >
    <div
      class="card-header"
    >
      <img
        alt="cute cat"
        class="card-img"
        src="https://images.unsplash.com/photo-1548247416-ec66f4900b2e?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8Mnx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=1000&q=60"
      />
      <button
        class="heart"
      >
        <img
          alt="outlined heart"
          src="heartOutlined.svg"
        />
      </button>
    </div>
    <div
      class="card-content"
    >
      <h3>
        Sydney
      </h3>
      <p>
        111-111-1111
      </p>
      <p>
        laith@hotmail.com
      </p>
      <p />
    </div>
  </article>
</DocumentFragment>
`;
